proposed/2021/DotNetAudit.md

```diff
@@ -0,0 +1,296 @@
+<!-- What is the impact of not doing this? -->
+There already exists a dotnet CLI commands called `dotnet list package` with options such as `--vulnerable, --deprecated, --outdated` which will list any known vulnerabilities, deprecations, and outdated packages in a project or solution. These options currently cannot be combined. Although this provides valuable information to understand the state of your dependencies, there does not exist a tool that allows you to quickly audit a project/solution & provide a way to act further on.
+
+When a developer is prompted with information such as a known vulnerability, deprecated package, or outdated version, they should have a clear understanding of how to best proceed with the information provided. In most cases, simply updating the dependency will suffice.
+
+Additionally, there already exists many third-party solutions that try to solve this problem with varying degrees of success. Some solutions can help alert about known vulnerabilities and some can help alert about outdated packages. No single tool seems to check for a "healthy" dependency list & that's another reason why we're looking to combine these experiences into a single command.
+
+## Prior Art
+
+<!-- What prior art, both good and bad are related to this proposal? -->
+<!-- Do other features exist in other ecosystems and what experience have their community had? -->
+<!-- What lessons from other communities can we learn from? -->
+<!-- Are there any resources that are relevent to this proposal? -->
+- [snyk](https://snyk.io/)
+- [npm audit](https://docs.npmjs.com/cli/v7/commands/npm-audit)
+- [cargo audit](https://github.com/RustSec/cargo-audit)
+- [dotnet outdated](https://github.com/dotnet-outdated/dotnet-outdated)
+- [dotnet retire](https://github.com/retirenet/dotnet-retire)
+- [NuGet Defense](https://github.com/digitalcoyote/NuGetDefense)
+  
+## Unresolved Questions
+
+<!-- What parts of the proposal do you expect to resolve before this gets accepted? -->
+<!-- What parts of the proposal need to be resolved before the proposal is stabilized? -->
+<!-- What related issues would you consider out of scope for this proposal but can be addressed in the future? -->
+- Should the command be named `audit` or `check`?
+  - `audit` is the more consistent name for package manager tooling with other ecosystems.
+  - `check` is the more consistent name with dotnet CLI.
+- Should this command only audit `vulnerabilities`? Or should it audit & be proactive as to `vulnerabilities`, `deprecations`, and `outdated` packages?
+- How much information should be present in the --json output?
+
+## Future Possibilities
+
+<!-- What future possibilities can you think of that this proposal would help with? -->
+- `dotnet audit` can be run on every `restore` which can throw warnings or errors to the user to take action against a vulnerable, deprecated, or outdated software supply chain.
```

Comment:
Could customer suppress warnings/errors?  It could be low impact noise. Also, even package has vulnerability and there is no alternative then customer stuck with it and took other mitigation steps (isolate network from outside world), then keep throwing error could discourage further usage.

Reply:
Yes I think we can figure out something. I am not so sure what arguments that might be right now as this experience will require an internet connection or a local cache. It might be something like `--offline`, `--level high`, or `--features vulnerability`.